data/transactions/logic/assembler.go

```diff
@@ -1155,81 +1211,99 @@ func (ops *OpStream) Bytes() (program []byte, err error) {
-	for o, l := range ops.offsetToLine {
+	newOffsetToLine := make(map[int]int, len(ops.OffsetToLine))
+	for o, l := range ops.OffsetToLine {
 		newOffsetToLine[o+pbl] = l
 	}
-	ops.offsetToLine = newOffsetToLine
+	ops.OffsetToLine = newOffsetToLine
 
-	program = out
-	return
+	return out
 }
 
 func (ops *OpStream) error(problem interface{}) error {
+	var le *lineError
 	switch p := problem.(type) {
 	case string:
-		return &lineError{Line: ops.sourceLine, Err: errors.New(p)}
+		le = &lineError{Line: ops.sourceLine, Err: errors.New(p)}
 	case error:
-		return &lineError{Line: ops.sourceLine, Err: p}
+		le = &lineError{Line: ops.sourceLine, Err: p}
 	default:
-		return &lineError{Line: ops.sourceLine, Err: fmt.Errorf("%#v", p)}
+		le = &lineError{Line: ops.sourceLine, Err: fmt.Errorf("%#v", p)}
 	}
+	ops.Errors = append(ops.Errors, le)
+	return le
 }
 
 func (ops *OpStream) errorf(format string, a ...interface{}) error {
 	return ops.error(fmt.Errorf(format, a...))
 }
 
-// AssembleString takes an entire program in a string and assembles it to bytecode using AssemblerDefaultVersion
-func AssembleString(text string) ([]byte, error) {
-	return AssembleStringWithVersion(text, assemblerNoVersion)
+func (ops *OpStream) warn(problem interface{}) error {
+	var le *lineError
+	switch p := problem.(type) {
+	case string:
+		le = &lineError{Line: ops.sourceLine, Err: errors.New(p)}
+	case error:
+		le = &lineError{Line: ops.sourceLine, Err: p}
+	default:
+		le = &lineError{Line: ops.sourceLine, Err: fmt.Errorf("%#v", p)}
+	}
+	warning := fmt.Errorf("warning: %w", le)
+	ops.Warnings = append(ops.Warnings, warning)
+	return warning
 }
-
-// AssembleStringV1 takes an entire program in a string and assembles it to bytecode using TEAL v1
-func AssembleStringV1(text string) ([]byte, error) {
-	return AssembleStringWithVersion(text, 1)
+func (ops *OpStream) warnf(format string, a ...interface{}) error {
+	return ops.warn(fmt.Errorf(format, a...))
 }
 
-// AssembleStringV2 takes an entire program in a string and assembles it to bytecode using TEAL v2
-func AssembleStringV2(text string) ([]byte, error) {
-	return AssembleStringWithVersion(text, 2)
+// ReportProblems issues accumulated warnings and errors to stderr.
+func (ops *OpStream) ReportProblems(fname string) {
+	for i, e := range ops.Errors {
+		if i > 9 {
+			break
+		}
+		fmt.Fprintf(os.Stderr, "%s: %s\n", fname, e)
+	}
+	for i, w := range ops.Warnings {
+		if i > 9 {
+			break
+		}
+		fmt.Fprintf(os.Stderr, "%s: %s\n", fname, w)
+	}
 }
 
-// AssembleStringWithVersion takes an entire program in a string and assembles it to bytecode using the assembler version specified
-func AssembleStringWithVersion(text string, version uint64) ([]byte, error) {
-	program, _, err := AssembleStringWithVersionEx(text, version)
-	return program, err
+// AssembleString takes an entire program in a string and assembles it to bytecode using AssemblerDefaultVersion
+func AssembleString(text string) (*OpStream, error) {
```

Comment:
What third party tools do we need to think about?  Is there anyone linking with our modules?

As it was, the interface was already returning three things - the program bytes, the offset map, and err.  That was already a bit odd - the offset map was only used in tests.  Adding teal errors and warnings to the return signature would have made the function return five values which seems extremely unwieldy.  By combining the related ideas into a single abstraction, the OpStream, we keep related things together.

Reply:
I'm willing the rename the OpStream to "Assembly" if that makes sense.  I think it would be a better name for what is now.  It might even bring in the PragmaStream functionality.

Reply:
>What third party tools do we need to think about? Is there anyone linking with our modules?

I'm not aware of any but who knows.

> Adding teal errors and warnings to the return signature would have made the function return five values which seems extremely unwieldy.

Not adding things to return but package everything into err. Agree, this does not work for warnings.